### PR TITLE
rust: Require `native: true` for `proc-macro` targets and allow linki…

### PR DESCRIPTION
**Pull request by sdroege**
…ng into targets for other platforms

proc-macros are basically compiler extensions and must be compiled for the build machine so that the compiler can load and execute it at runtime.

Without this, cross-compilation of targets that use proc-macro dependencies is going to fail.

Fixes https://github.com/mesonbuild/meson/issues/11702

### CODE REVIEW

File: test cases/rust/18 proc-macro/meson.build
Comment (sdroege) [inline test cases/rust/18 proc-macro/meson.build:8]: A couple of lines above, why is this test disabled on non-Linux, how does it fail? Theoretically there shouldn't be anything special required.

File: mesonbuild/build.py
Comment (eli-schwartz) [inline mesonbuild/build.py:2332]: So it looks like this would prospectively error out on projects that have never attempted to support cross building, and are currently being compiled natively, because they are doing something technically wrong despite that it still compiles okay on native-only builds. Am I understanding correctly?

Comment (sdroege) [reply mesonbuild/build.py:2332]: Yes. Would you prefer a warning instead?

Comment (sdroege) [reply mesonbuild/build.py:2332]: I guess that would be better, for backwards compat. Even if the current way is just wrong :)

Comment (eli-schwartz) [reply mesonbuild/build.py]: I think that would be better, yes.

Keeping in mind that these projects cannot be cross compiled at *all* on older versions of Meson, so people were unlikely to bother setting that either way...

Whatever else, we want that existing builds which worked, should continue to work due to stability of the semver major release version.

Comment (sdroege) [reply mesonbuild/build.py]: Changed accordingly

Comment (dcbaker) [reply mesonbuild/build.py]: I think we could do something like:
```python
if kwargs.get('native', False):
    if self.environment.is_cross_build():
        error(...)
    else:
        warning(...)
```
since compilation would have already failed in the cross build case anyway, but would succeed otherwise?

Comment (sdroege) [reply mesonbuild/build.py]: Sounds correct, yes. That would also work for me, your call :)

**Comment by jpakkane** [comment]
Is there ever a case where you'd want to compile `proc-macro` crates with a cross compiler? If no, could we not make it default to `native: true` instead of requiring people to write it out by hand every time?

File: test cases/rust/18 proc-macro/meson.build
Comment (tristan957) [reply test cases/rust/18 proc-macro/meson.build:8]: @sdroege some of the tests are disabled on certain platforms because toolchains might not be setup on those platforms (work needed) or tests can take too long. In this case, you are probably running into the second option. Azure CI takes an ungodly amount of time and will timeout frequently.

File: test cases/rust/18 proc-macro/meson.build
Comment (eli-schwartz) [reply test cases/rust/18 proc-macro/meson.build:11]: See e.g. cross/linux-mingw-w64-64bit.txt

You just need to set up the rust compiler as a cross binary, and configure the corresponding .json file to change e.g. `"tests": ["common", "cmake"],` to specify which run_project_tests.py suites are now supported by that cross test configuration.

**Comment by sdroege** [comment]
> Is there ever a case where you'd want to compile `proc-macro` crates with a cross compiler? If no, could we not make it default to `native: true` instead of requiring people to write it out by hand every time?

That would also work, I guess.

The only case I could think of is when someone wants to compile the proc-macro on some build machine, and then compile something *with* that proc-macro on the target machine. Not sure if that's something worth worrying about?

**Comment by eli-schwartz** [comment]
You'd have to install one to do that. Is this expected to work?

**Comment by sdroege** [comment]
You'd have to install it or copy it around somehow manually, yes. I wouldn't expect that to work but with enough motivation and carefully set up environments you can probably make it work. It's something I wouldn't want to support though.

**Comment by eli-schwartz** [comment]
Ping. What's the status of wiring up tests?

**Review by dcbaker — APPROVED** [review]
From a code point of view this looks fine to me, I still wonder long-term if we want to deprecate the proc-macro crate type and just add  a `rust_module.proc_macro()` helper to avoid this. I think this is a net improvement over what we have, so I'm happy

**Comment by dcbaker** [comment]
I'm also in favor of landing this after the rc if necessary. From my point of view this is a bug fix, since while Meson currently allows something that cannot work in practice

**Comment by eli-schwartz** [comment]
Agreed -- once it is tested in CI I'm happy with merging it even past the rc freeze.

**Comment by sdroege** [comment]
> What's the status of wiring up tests?

Doing tests looks like an adventure that will take me a couple of hours once I figured out how the CI works exactly and waited for a few dozen failed runs in the meantime, so I didn't get to that yet and it will probably take a while. I'd like to have this merged but I also think that having tests on the CI for this is important because otherwise it will just stop working again sooner or later :)

If someone wants to take this over please go ahead, otherwise I'll get to it once I have a bigger slot of time available.

File: mesonbuild/build.py
Comment (xclaesse) [inline mesonbuild/build.py:1411]: I think `t` can be a `CustomTarget` which does not implement `uses_rust()`. What about:

```suggestion
            if self.for_machine is not t.for_machine and (not isinstance(t, SharedLibrary) or t.rust_crate_type != 'proc-macro'):
```

File: mesonbuild/build.py
Comment (xclaesse) [inline mesonbuild/build.py:2337]: It does not require "native: true", that's the default in this case. I think it should always be an hard error, no need to downgrade to a warning if it's not a cross build. To happen, user must have explicitely wrote:
```meson
shared_library(..., rust_crate_type: 'proc-macros', native: false)
```
which is always wrong.

What about:
```suggestion
                    raise InvalidArguments('Rust "proc-macro" crate type does not allow "native: false".')
```

File: test cases/rust/18 proc-macro/meson.build
Comment (xclaesse) [reply test cases/rust/18 proc-macro/meson.build:11]: I don't think we should add `native: true` here, the whole point of this PR is to test it's the default.

Giving it a try at cross rust compilation in CI there: https://github.com/mesonbuild/meson/pull/11990.

**Comment by heftig** [comment]
I'm encountering a bit more trouble cross-compiling Mesa for i686; the `rusticl_proc-macros` shared lib is used to build the `rusticl` static lib, which is then `link_whole`d into the `RusticlOpenCL` shared lib.

Meson adds the `rusticl_proc-macros` shared lib to the link args of the `RusticlOpenCL` shared lib, which doesn't do anything in a native build but fails to link in a cross build.

File: test cases/rust/18 proc-macro/meson.build
Comment (xclaesse) [reply test cases/rust/18 proc-macro/meson.build:8]: It could very well be historical reasons that have been fixed since then. Feel free to open a separate PR that unskip them and see how it goes on the CI.

**Comment by xclaesse** [comment]
@heftig Can you please open a separate issue? A minimal test case would be nice too because I failed to reproduce.

**Comment by xclaesse** [comment]
Closing this PR because it is included into https://github.com/mesonbuild/meson/pull/11714 which does a bigger refactoring of proc-macro.

**Pull request closed without merging**
